### ISSUE TITLE
Contribution is not handled when it is made from a webform.

### ISSUE THREAD

**Issue by jitendrapurohit** [agileware/au.com.agileware.ewayrecurring#4]
We are using the updated code of this ext for one of our clients - and it seems the eway response is only handled for a contribution page and not when the payment is made through webforms? See https://github.com/agileware/au.com.agileware.ewayrecurring/blob/master/eWAYRecurring.php#L278

I've looked into the code to fix this and handled the response in a custom module using a webform hook that is called at the end of the process similar to what is processed by the ext when a  civicrm thankyou page is displayed.

    /**
     * Handle eway return on webform.
    */
    function hook_preprocess_webform_confirmation(&$vars) {
      civicrm_initialize();
      $sid = CRM_Utils_Array::value('sid', $_GET);
      $accessCode = CRM_Utils_Array::value('AccessCode', $_GET);
      if (!empty($sid) && !empty($accessCode)) {
        module_load_include('inc', 'webform', 'includes/webform.submissions');
        $submission = webform_get_submissions(array('sid' => $sid));
        $contributionID = $submission[$sid]->civicrm['contribution'][1]['id'];
        $contribution = civicrm_api3('Contribution', 'get', [
          'sequential' => 1,
          'id' => $contributionID,
          'sequential' => TRUE,
          'return'     => array('contribution_page_id', 'contribution_recur_id', 'is_test'),
        ]);

        if (count($contribution['values']) > 0) {
          $res = CRM_Core_Resources::singleton();
          $ewayPath = $res->getPath('au.com.agileware.ewayrecurring');
          // Include eWay SDK.
          require_once $ewayPath . '/vendor/autoload.php';

          $contribution = $contribution['values'][0];
          $pp = CRM_Financial_BAO_PaymentProcessor::getProcessorForEntity($contributionID);
          $def = array();
          $par = array('id' => $pp);
          $ppObj = (array) CRM_Financial_BAO_PaymentProcessor::retrieve($par, $def);
          $paymentProcessor = new au_com_agileware_ewayrecurring('live', $ppObj);
          $paymentProcessor->validateContribution($accessCode, $contribution, $qfKey);
        }
      }
    }

The above is working fine for us and do complete the transaction after returning from the eway site. Do you agree something similar should be handled within the ext to also support webform payments?

Also, I did some changes to the request params that we send to eway, eg `country_id` instead of `country` at https://github.com/agileware/au.com.agileware.ewayrecurring/blob/master/au_com_agileware_ewayrecurring.class.php#L302 and converting it into an iso code as done below that line.

**Comment by jusfreeman** [comment]
Hi @jitendrapurohit - as discussed with @petednz today, the master branch has very limited testing with Drupal Webform CiviCRM. Thanks for the feedback and contribution.

**Comment by agilewarealok** [comment]
Hi @jitendrapurohit 

Thanks for your feedback and possible solution. 

We've tried to come up with a different solution, 
We've created a Schedule Job which will verify the pending eWay transactions. This will be useful in any case when CiviCRM is unable to verify the payment just after the user navigates back to the site.

Changes are pushed under https://github.com/agileware/au.com.agileware.ewayrecurring/tree/CIVIWF-1-Queue-2.0
